File: AoC.2023.FSharp/Day23.fs
module AoC._2023.FSharp.Day23

open AoC.FSharp
open AoC.Library.Runner
open AoC.Library.Utils

type Point = PointBase<int>

type Cell =
    | Wall
    | Slope of Direction
    | Empty

let mapCell =
    function
    | '#' -> Wall
    | '>' -> Slope(Direction.Right)
    | 'v' -> Slope(Direction.Down)
    | '<' -> Slope(Direction.Left)
    | _ -> Empty

let neig map x y p2 =
    Array2D.get map x y
    |> function
        | Slope(dir) when not p2 -> Seq.singleton dir
        | Wall -> Seq.empty
        | _ -> Seq.ofList [ Direction.Up; Direction.Right; Direction.Down; Direction.Left ]
    |> Seq.map Point.op_Implicit
    |> Seq.map (fun p -> p.X + x, p.Y + y)
    |> List.ofSeq

let search map x0 y0 p2 =
    let w = Array2D.length1 map
    let h = Array2D.length2 map

    let rec searchInn =
        function
        | [] -> 0
        | (x, y, _) :: tail when x < 0 || y < 0 || x >= w && y >= h -> searchInn tail
        | (x, y, v) :: tail when List.contains (x, y) v -> searchInn tail
        | (x, y, visits) :: tail ->
            match Array2D.get map x y with
            | Wall -> searchInn tail
            | _ when y = h - 1 ->  
                let curr = List.length visits
                max curr (searchInn tail)
            | _ ->
                let newVisits = visits @ [x, y]
                let neighbors = neig map x y p2
                let q = tail @ (List.map (function (x, y) -> (x, y, newVisits)) neighbors)
                    
                searchInn q

    searchInn [ x0, y0, [] ]



[<DateInfo(2023, 23, AdventParts.PartTwo)>]
type Day23() =
    inherit AdventSolution()

    override this.SolvePartOne() =
        let map = this.Input.SquareMap() |> Array2D.map mapCell

        let startX =
            map
            |> Array2D.toSeq
            |> Seq.indexed
            |> Seq.where (snd >> (<>) Wall)
            |> Seq.head
            |> fst

        this.WriteLine $"Starting at {startX}"

        search map startX 0 false

    override this.SolvePartTwo() =
        let map = this.Input.SquareMap() |> Array2D.map mapCell

        let startX =
            map
            |> Array2D.toSeq
            |> Seq.indexed
            |> Seq.where (snd >> (<>) Wall)
            |> Seq.head
            |> fst

        this.WriteLine $"Starting at {startX}"

        search map startX 0 true

File: AoC.2023/Day23.cs
using System.Runtime.CompilerServices;
using AoC.Library.Runner;
using AoC.Library.Utils;

namespace AoC._2023;

using Point = PointBase<int>;

[DateInfo(2023, 23, AdventParts.All)]
public class Day23 : AdventSolution
{
    private static readonly Direction[] Dirs = {
        Direction.Right, Direction.Down, Direction.Up, Direction.Left
    };

    private int _currMax = 0;
    private bool[,] _visits = null!;

    private readonly Stack<Point> _path = new();

    private int Solve(int startX, bool isPartOne)
    {
        _visits = Input.SquareMap(_ => false);

        return InnerSolve(new Point(startX, 0), isPartOne) - 1;
    }

    private int InnerSolve(Point start, bool isPartOne)
    {
        var pushed = 1;
        _path.Push(start);

        while (true)
        {
            var curr = _path.Peek();

            if (curr.Y == Input.Height - 1)
            {
                var res = _path.Count;

                if (res > _currMax)
                {
                    Console.WriteLine(_currMax = res);
                }

                Clean(pushed);

                return res;
            }

            _visits[curr.X, curr.Y] = true;

            var neig = (Input[curr] switch {
                    '#' => Array.Empty<Direction>(),
                    '>' when isPartOne => Direction.Right.Single(),
                    'v' when isPartOne => Direction.Down.Single(),
                    '<' when isPartOne => Direction.Left.Single(),
                    _ => Dirs
                })
                .Select(d => d + curr)
                .Where(p => p.InBounds(Input.Width, Input.Height))
                .Where(p => Input[p] != '#')
                .Where(p => !_visits[p.X, p.Y])
                .ToArray();

            switch (neig.Length)
            {
                case 0:
                    Clean(pushed);

                    return 0;

                case 1:
                    _path.Push(neig[0]);
                    pushed++;

                    break;

                default:
                    var currMax = 0;

                    foreach (var n in neig)
                    {
                        currMax = Math.Max(InnerSolve(n, isPartOne), currMax);
                    }

                    Clean(pushed);

                    return currMax;
            }
        }
    }

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private void Clean(int pushed)
    {
        for (var i = 0; i < pushed; i++)
        {
            var curr = _path.Pop();
            _visits[curr.X, curr.Y] = false;
        }
    }

    public override object SolvePartOne()
    {
        var startX = Input[0].IndexOf('.');


        return Solve(startX, true);
    }

    public override object SolvePartTwo()
    {
        var startX = Input[0].IndexOf('.');

        return Solve(startX, false);
        // launched at 1:08
        //6235 01:10 1:41
        //6491 01:12
    }
}